src/commands/protocols/publish/read-cached-publish-data-command.js

```diff
@@ -0,0 +1,106 @@
+import { setTimeout } from 'timers/promises';
+import Command from '../../command.js';
+import {
+    OPERATION_ID_STATUS,
+    ERROR_TYPE,
+    MAX_RETRIES_READ_CACHED_PUBLISH_DATA,
+    RETRY_DELAY_READ_CACHED_PUBLISH_DATA,
+} from '../../../constants/constants.js';
+
+class ReadCachedPublishDataCommand extends Command {
+    constructor(ctx) {
+        super(ctx);
+        this.ualService = ctx.ualService;
+        this.dataService = ctx.dataService;
+        this.fileService = ctx.fileService;
+        this.repositoryModuleManager = ctx.repositoryModuleManager;
+        this.networkModuleManager = ctx.networkModuleManager;
+
+        this.errorType = ERROR_TYPE.STORE_ASSERTION_ERROR;
+    }
+
+    async execute(command) {
+        const { event } = command.data;
+        const eventData = JSON.parse(event.data);
+        const { id, publishOperationId, merkleRoot, chunksAmount } = eventData;
+        const { blockchain, contractAddress } = event;
+        const operationId = await this.operationIdService.generateOperationId(
+            OPERATION_ID_STATUS.PUBLISH_FINALIZATION.PUBLISH_FINALIZATION_START,
+            publishOperationId,
+        );
+        let datasetPath;
+        let cachedData;
+
+        for (let attempt = 1; attempt <= MAX_RETRIES_READ_CACHED_PUBLISH_DATA; attempt += 1) {
```

Comment:
Is there no way to use the command retry logic instead ? In general, it's preferable, as we don't want command queue to get blocked

Reply:
You can set delay and number of retries in command data

Reply:
the thing is, we actually want this to be blocking - because the blockchain event can get emitted before the cached data is available, and only then to execute this command's logic and the rest of the command sequence (tho these constants were purely for testing, waiting for someone to review it so we know what exactly to put as the retry delay and amounts cc @Mihajlo-Pavlovic )

ill check again tomorrow with a fresh brain whether this can be done with command retry logic, but anyways our plan was to go with this approach since its quick and make it cleaner later. But since u have a bunch of experience, if u think this will be a major problem on testnet/mainnet let us know so we know what to expect

Reply:
What I meant is you can:
1. schedule command on event
2. read from file
3. if data is missing, schedule retry of command. If it's there, continue with rest of logic.

This looks to me like it's the same thing as you're doing now, but the node is not waiting during command execution, it's just rescheduling at a later point in time.

It's hard for me to tell if this is a problem or not. I'm not sure how often this command is scheduled, and how many of these in parallel are executed...

Reply:
pretty sure you're right, will check w rest of the team and do it that way if it works the way we want to